File: algorithms/3_hard/M/01520_maximum_number_of_non_overlapping_substrings/solution.cpp
constexpr size_t INF = numeric_limits<size_t>::max();

class Solution {
public:
    static vector<string> maxNumOfSubstrings(const string s) {
        if (s.empty()) { return vector<string>(); }

        const size_t len = s.size();
        size_t seen[26][4] = { 0 };
        for (size_t i = 0; i < 26; i += 1) {
            seen[i][0] = INF; // first_seen
            seen[i][1] = INF; // last_seen
            seen[i][2] = 0;   // mask@last
            seen[i][3] = 0;   // running_mask
        }

        for (size_t i = 0; i < len; i += 1) {
            const size_t ch_idx = (size_t)(s[i] - 'a');
            if (seen[ch_idx][0] == INF) {
                seen[ch_idx][0] = i;
            }

            const size_t mask = 1 << ch_idx;
            for (size_t j = 0; j < 26; j += 1) {
                if (seen[j][0] != INF) {
                    seen[j][3] |= mask;
                }
            }
            seen[ch_idx][1] = i;
            seen[ch_idx][2] = seen[ch_idx][3];
        }

        // Almost Bellman-Ford idea
        for (size_t steps = 0; steps < 26; steps += 1) {
            bool has_update = false;
            for (size_t i = 0; i < 26; i += 1){
                if (seen[i][0] != INF) {
                    for (size_t j = 0; j < 26; j += 1) {
                        const size_t mask_j = 1 << j;
                        if ((seen[i][2] & mask_j) != 0) {
                            if (seen[j][0] < seen[i][0]) {
                                has_update = true;
                                seen[i][0] = seen[j][0];
                            }

                            if (seen[j][1] > seen[i][1]) {
                                has_update = true;
                                seen[i][1] = seen[j][1];
                            }
                        }
                    }
                }
            }

            if (!has_update) {
                break;
            }
        }

        auto res = vector<string>();
        auto valids = vector<pair<size_t, size_t>>();
        for (size_t i = 0; i < 26; i += 1) {
            if (seen[i][0] != INF) {
                valids.push_back({ seen[i][0], seen[i][1] });
            }
        }
        ranges::sort(valids, [](const auto a, const auto b) {
            return a.second < b.second;
        });

        res.push_back(s.substr(valids[0].first, valids[0].second - valids[0].first + 1));

        size_t prev_end = valids[0].second;
        for (const auto [lo, hi] : valids) {
            if (lo > prev_end) {
                prev_end = hi;
                res.push_back(s.substr(lo, hi - lo + 1));
            }
        }

        return res;
    }
};

File: algorithms/3_hard/M/01520_maximum_number_of_non_overlapping_substrings/solution.rs
impl Solution {
    const INF: usize = usize::MAX;

    pub fn max_num_of_substrings(s: String) -> Vec<String> {
        if s.is_empty() { return vec![]; }

        //               first      last   mask@last running_mask
        let mut seen = [(Self::INF, Self::INF, 0_u32, 0_u32); 26];
        let bytes = s.as_bytes();

        for (i, &b) in bytes.iter().enumerate() {
            debug_assert!(b >= b'a' && b <= b'z');
            let b_idx = (b - b'a') as usize;
            if seen[b_idx].0 == Self::INF {
                seen[b_idx].0 = i;
            }

            let mask = 1_u32 << (b_idx as u32);
            for j in 0..26 {
                if seen[j].0 != Self::INF {
                    seen[j].3 |= mask;
                }
            }
            seen[b_idx].1 = i;
            seen[b_idx].2 = seen[b_idx].3;
        }

        // Almost Bellman-Ford idea
        for _ in 0..26 {
            let mut has_update = false;
            for i in 0..26 {
                if seen[i].0 != Self::INF {
                    for j in 0..26 {
                        let mask_j = 1_u32 << (j as u32);
                        if (seen[i].2 & mask_j) != 0 {
                            if seen[j].0 < seen[i].0 {
                                has_update = true;
                                seen[i].0 = seen[j].0;
                            }

                            if seen[j].1 > seen[i].1 {
                                has_update = true;
                                seen[i].1 = seen[j].1;
                            }
                        }
                    }
                }
            }

            if !has_update {
                break;
            }
        }

        let mut res: Vec<String> = Vec::new();
        let mut valids: Vec<(usize, usize)> =  seen.iter().filter(|x| x.0 != Self::INF).map(|x| (x.0, x.1)).collect::<Vec<_>>();
        valids.sort_unstable_by_key(|x| x.1);
        res.push(String::from_utf8_lossy(&bytes[valids[0].0..=valids[0].1]).into_owned());

        let mut prev_end = valids[0].1;
        for (lo, hi) in valids.into_iter().skip(1) {
            if lo > prev_end {
                prev_end = hi;
                res.push(String::from_utf8_lossy(&bytes[lo..=hi]).into_owned());
            }
        }

        res
    }
}
